fuzz: Make p2p_headers_presync more deterministic

maflcko opened the pull request:
This should make the `p2p_headers_presync` fuzz target more deterministic.

Tracking issue: https://github.com/bitcoin/bitcoin/issues/29018.

The first commits adds an `ElapseSteady` helper and type aliases. The second commit uses those helpers in `ReportHeadersPresync` and in the fuzz target to increase determinism.

### Testing

It can be tested via (setting 32 parallel threads):

```
cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../b-c-qa-assets/fuzz_corpora/ p2p_headers_presync 32
```

The failing diff is contained in the commit messages, if applicable.

Commit:
test: Introduce MockableSteadyClock::mock_time_point and ElapseSteady helper

This refactor clarifies that the MockableSteadyClock::mock_time_point
has millisecond precision by defining a type an using it.

Moreover, a ElapseSteady helper is added which can be re-used easily.

Commit:
refactor: Use MockableSteadyClock in ReportHeadersPresync

This allows the clock to be mockable in tests. Also, replace cs_main
with GetMutex() while touching this function.

Also, use the ElapseSteady test helper in the p2p_headers_presync fuzz
target to make it more deterministic.

The m_last_presync_update variable is a global that is not reset in
ResetAndInitialize. However, it is only used for logging, so completely
disable it for now.

Without this patch, the tool would report a diff:

cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../qa-assets/fuzz_corpora/ p2p_headers_presync 32

...
  4468|     81|        auto now = std::chrono::steady_clock::now();
  4469|     81|        if (now < m_last_presync_update + std::chrono::milliseconds{250}) return;
-                                                                                        ^80
+                                                                                        ^79
...

Commit:
fuzz: Avoid setting the mock-time twice

It should be sufficient to set it once. Especially, if the dynamic value
is only used by ResetAndInitialize.

This also avoids non-determistic code paths, when ResetAndInitialize may
re-initialize m_next_inv_to_inbounds.

Without this patch, the tool would report a diff:

cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../qa-assets/fuzz_corpora/ p2p_headers_presync 32

...
- 1126|      3|        m_next_inv_to_inbounds = now + m_rng.rand_exp_duration(average_interval);
- 1127|      3|    }
+ 1126|     10|        m_next_inv_to_inbounds = now + m_rng.rand_exp_duration(average_interval);
+ 1127|     10|    }
  1128|    491|    return m_next_inv_to_inbounds;
...

Commit:
fuzz: Set ignore_incoming_txs in p2p_headers_presync

This avoids non-determistic code paths.

Without this patch, the tool would report a diff:

cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../qa-assets/fuzz_corpora/ p2p_headers_presync 32

...
- 5371|    393|        peer.m_next_send_feefilter = current_time + m_rng.randrange<std::chrono::microseconds>(MAX_FEEFILTER_CHANGE_DELAY);
- 5372|    393|    }
+ 5371|    396|        peer.m_next_send_feefilter = current_time + m_rng.randrange<std::chrono::microseconds>(MAX_FEEFILTER_CHANGE_DELAY);
+ 5372|    396|    }
  5373|  16.2k|}
...

marcofleon commented:
> Are you using libFuzzer?

yeah just figured this out, disregard. Do you know why this is? What about libFuzzer trips it up?

marcofleon commented:
@dergoegge and I checked it out. Turns out libfuzzer shuffles the corpus, so running it twice with a different ordering of inputs can result in different coverage. On this PR, running with libFuzzer and setting `-shuffle=0` should produce no coverage differences for `p2p_headers_presync`. 

I think this is another sign of non-determinism, as the ordering shouldn't affect line hit counts. It seems that using libFuzzer does end up providing a bit of new information.

Crypt-iQ commented:
This may be a shot in the dark, but is it possible that `ResetAndInitialize` re-initializing the `id` counter to 0 introduces some non-determinism across runs depending on the ordering of the corpus? During normal bitcoind operation, the `NodeId` is always monotonically increasing; is there a reason it is reset in `p2p_headers_presync` and `process_messages`?

dergoegge commented:
> However, I am not seeing the global state that causes this. For reference, one part of the non-determinism is fixed by the change in commit https://github.com/bitcoin/bitcoin/commit/2222dd8e60d82e5aa96f3d80a81ddca880d627ea. I am happy to re-add the commit here, but I'd feel better if I fully understood the commit myself.

I think it might be the `m_rng` on peerman which is a global in this test.

In the following `m_rng` is only used if `tx_relay->m_next_inv_send_time < current_time` and https://github.com/bitcoin/bitcoin/commit/2222dd8e60d82e5aa96f3d80a81ddca880d627ea causes that to no longer be the case:
https://github.com/bitcoin/bitcoin/blob/65dcbec75661d1652beb927f03b1feab4fab932e/src/net_processing.cpp#L5679-L5686

> is there a reason it is reset in p2p_headers_presync and process_messages?

For each test case we want to simulate starting out with no peers (i.e. id = 0) and then add peers and send messages.

Crypt-iQ commented:
> For each test case we want to simulate starting out with no peers (i.e. id = 0) and then add peers and send messages.

That makes sense, I think my question was a bit unclear. There are two places I can see in net_processing where `NodeId` persists even after the peer has disconnected (i.e. polluting global state somewhat between fuzz iterations):
- `lNodesAnnouncingHeaderAndIDs`
- `m_headers_presync_bestpeer`

The first one does not have any fuzz coverage, but the second one is executed by `p2p_headers_presync`. Running the full command in the OP above gives the following result where `m_headers_presync_bestpeer` [no longer exists in `m_headers_presync_stats`](https://github.com/bitcoin/bitcoin/blob/65dcbec75661d1652beb927f03b1feab4fab932e/src/net_processing.cpp#L2587-L2599):
```diff
 diff --git a/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.a.txt b/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.b.txt
 index 0822ec1f73..74efd6c71d 100644
  --- a/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.a.txt
  +++ b/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.b.txt
  @@ -102156,38 +102156,38 @@
    2586|  3.79k|            bool best_updated = false;
    2587|  3.79k|            if (best_it == m_headers_presync_stats.end()) {
     ------------------
  -  |  Branch (2587:17): [True: 37, False: 3.76k]
  +  |  Branch (2587:17): [True: 39, False: 3.75k]
     ------------------
    2588|       |                // If the cached best peer is outdated, iterate over all remaining ones (including
    2589|       |                // newly updated one) to find the best one.
  - 2590|     37|                NodeId peer_best{-1};
  - 2591|     37|                const HeadersPresyncStats* stat_best{nullptr};
  - 2592|     37|                for (const auto& [peer, stat] : m_headers_presync_stats) {
  + 2590|     39|                NodeId peer_best{-1};
  + 2591|     39|                const HeadersPresyncStats* stat_best{nullptr};
  + 2592|     39|                for (const auto& [peer, stat] : m_headers_presync_stats) {
     ------------------
  -  |  Branch (2592:47): [True: 37, False: 37]
  +  |  Branch (2592:47): [True: 39, False: 39]
     ------------------
  - 2593|     37|                    if (!stat_best || stat > *stat_best) {
  + 2593|     39|                    if (!stat_best || stat > *stat_best) {
                                                       ^0
     ------------------
  -  |  Branch (2593:25): [True: 37, False: 0]
  +  |  Branch (2593:25): [True: 39, False: 0]
     |  Branch (2593:39): [True: 0, False: 0]
     ------------------
  - 2594|     37|                        peer_best = peer;
  - 2595|     37|                        stat_best = &stat;
  - 2596|     37|                    }
  - 2597|     37|                }
  - 2598|     37|                m_headers_presync_bestpeer = peer_best;
  - 2599|     37|                best_updated = (peer_best == pfrom.GetId());
  - 2600|  3.76k|            } else if (best_it->first == pfrom.GetId() || stats > best_it->second) {
  + 2594|     39|                        peer_best = peer;
  + 2595|     39|                        stat_best = &stat;
  + 2596|     39|                    }
  + 2597|     39|                }
  + 2598|     39|                m_headers_presync_bestpeer = peer_best;
  + 2599|     39|                best_updated = (peer_best == pfrom.GetId());
  + 2600|  3.75k|            } else if (best_it->first == pfrom.GetId() || stats > best_it->second) {
```

Changing the `NodeId` to be a member of `HeadersSyncSetup` that is monotonically increasing does get rid of the above non-determinism, but it seems to introduce some new non-determinism elsewhere.

marcofleon commented:
Commit 2222dd8e60d82e5aa96f3d80a81ddca880d627ea added to this PR fixes all non-determinism for me. I should probably run the script more than 10 times to be sure, but so far so good.

@Crypt-iQ the part you're highlighting is the only code where the connection to Niklas's rationale isn't immediately clear. I think it could be in `MaybeSendPing` where sometimes a peer gets marked for disconnection based on some timeout which is dependent on mock time. That node then gets removed from `m_header_presync_stats` in `FinalizeNode`. So having the two different mock times and a different input ordering would affect which peers are part of the presync.

Crypt-iQ commented:
@marcofleon That sounds plausible; I was thinking along the lines of the `NodeId` getting reused across iterations:
- Iteration 1: `id=0` is set as `m_headers_presync_bestpeer`
- Iteration 2: `ResetAndInitialize` calls `FinalizeNode` which wipes all peer state except for `m_headers_presync_bestpeer`
- Iteration 2: Depending on the corpus input, the peer with `id=0` sends a `HEADERS` message which eventually calls `IsContinuationOfLowWorkHeadersSync`. This will [populate](https://github.com/bitcoin/bitcoin/blob/65dcbec75661d1652beb927f03b1feab4fab932e/src/net_processing.cpp#L2584-L2585) the `m_headers_presync_stats` with `id=0` which then means that the best peer will be found in `m_headers_presync_stats`.

Commit:
fuzz: Shuffle files before testing them

When iterating over all fuzz input files in a folder, the order should
not matter.

However, shuffling may be useful to detect non-determinism.

Thus, shuffle in fuzz.cpp, when using neither libFuzzer, nor AFL.

Also, shuffle in the deterministic-fuzz-coverage tool, when using
libFuzzer.

marcofleon commented:
That's a good explanation, makes sense to me. I spent some time trying to connect what you're saying to the mock time fix, but turns out there is no connection. I forgot that taking out `SetMockTime(ConsumeTime(fuzzed_data_provider))` changes the input format so running on the old corpus isn't useful. As a workaround to generating a new corpus, I left the `ConsumeTime` in the test but didn't set it to mock time and this specific instability still shows up. It's separate from https://github.com/bitcoin/bitcoin/pull/32198#issuecomment-2778718945 then.

Is `m_header_presync_bestpeer` only used for reporting/logging to validation? I'm not sure how useful to the test it would be to fix it.

Commit 2222dd8e60d82e5aa96f3d80a81ddca880d627ea seems to address some of the other non-determinism that shows up so I think including it in this PR makes sense.

Commit:
fuzz: Move global node id counter along with other global state

The global m_headers_presync_stats is not reset in ResetAndInitialize.
This may lead to non-determinism.

Fix it by incrementing the global node id counter instead.

Without this patch, the tool would report a diff:

cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../qa-assets/fuzz_corpora/ p2p_headers_presync 32

...
  2587|  3.73k|            if (best_it == m_headers_presync_stats.end()) {
   ------------------
-  |  Branch (2587:17): [True: 80, False: 3.65k]
+  |  Branch (2587:17): [True: 73, False: 3.66k]
   ------------------
...

Crypt-iQ commented:
tACK fa2d3441c6557adba3d393abfc1af22fb788b5c2

Ran the script 100 times and it didn't give any errors.

> Let me know if this should be fixed.

I think this should be fixed to not give false positives, but I think this can be done in another PR?

marcofleon commented:
Using a new corpus, this is coming up at times:
<details>
<summary>diff --git a/./fuzzbuild/fuzz_det_cov.show.t9.a.txt b/./fuzzbuild/fuzz_det_cov.show.t9.b.txt</summary>

```
index 3cb0b6f082..9e79f23f38 100644
--- a/./fuzzbuild/fuzz_det_cov.show.t9.a.txt
+++ b/./fuzzbuild/fuzz_det_cov.show.t9.b.txt
@@ -168783,28 +168783,26 @@
    28|       |    // newTaskMutex is locked throughout this loop EXCEPT
    29|       |    // when the thread is waiting or when the user's function
    30|       |    // is called.
-   31|      2|    while (!shouldStop()) {
+   31|      1|    while (!shouldStop()) {
   ------------------
-  |  Branch (31:12): [True: 1, False: 1]
+  |  Branch (31:12): [True: 0, False: 1]
   ------------------
-   32|      1|        try {
-   33|      2|            while (!shouldStop() && taskQueue.empty()) {
-                                                  ^1
+   32|      0|        try {
+   33|      0|            while (!shouldStop() && taskQueue.empty()) {
   ------------------
-  |  Branch (33:20): [True: 1, False: 1]
-  |  Branch (33:37): [True: 1, False: 0]
+  |  Branch (33:20): [True: 0, False: 0]
+  |  Branch (33:37): [True: 0, False: 0]
   ------------------
    34|       |                // Wait until there is something to do.
-   35|      1|                newTaskScheduled.wait(lock);
-   36|      1|            }
+   35|      0|                newTaskScheduled.wait(lock);
+   36|      0|            }
    37|       |
    38|       |            // Wait until either there is a new task, or until
    39|       |            // the time of the first item on the queue:
    40|       |
-   41|      1|            while (!shouldStop() && !taskQueue.empty()) {
-                                                  ^0
+   41|      0|            while (!shouldStop() && !taskQueue.empty()) {
   ------------------
-  |  Branch (41:20): [True: 0, False: 1]
+  |  Branch (41:20): [True: 0, False: 0]
   |  Branch (41:37): [True: 0, False: 0]
   ------------------
    42|      0|                std::chrono::steady_clock::time_point timeToWaitFor = taskQueue.begin()->first;
@@ -168818,40 +168816,39 @@
    47|       |
    48|       |            // If there are multiple threads, the queue can empty while we're waiting (another
    49|       |            // thread may service the task we were waiting on).
-   50|      1|            if (shouldStop() || taskQueue.empty())
-                                              ^0
+   50|      0|            if (shouldStop() || taskQueue.empty())
   ------------------
-  |  Branch (50:17): [True: 1, False: 0]
+  |  Branch (50:17): [True: 0, False: 0]
   |  Branch (50:33): [True: 0, False: 0]
   ------------------
    51|      1|                continue;
    52|       |
-   53|      0|            Function f = taskQueue.begin()->second;
-   54|      0|            taskQueue.erase(taskQueue.begin());
+   53|  18.4E|            Function f = taskQueue.begin()->second;
+   54|  18.4E|            taskQueue.erase(taskQueue.begin());
    55|       |
-   56|      0|            {
+   56|  18.4E|            {
    57|       |                // Unlock before calling f, so it can reschedule itself or another task
    58|       |                // without deadlocking:
-   59|      0|                REVERSE_LOCK(lock);
+   59|  18.4E|                REVERSE_LOCK(lock);
   ------------------
-  |  |  243|      0|#define REVERSE_LOCK(g) typename std::decay<decltype(g)>::type::reverse_lock UNIQUE_NAME(revlock)(g, #g, __FILE__, __LINE__)
+  |  |  243|  18.4E|#define REVERSE_LOCK(g) typename std::decay<decltype(g)>::type::reverse_lock UNIQUE_NAME(revlock)(g, #g, __FILE__, __LINE__)
   |  |  ------------------
-  |  |  |  |   11|      0|#define UNIQUE_NAME(name) PASTE2(name, __COUNTER__)
+  |  |  |  |   11|  18.4E|#define UNIQUE_NAME(name) PASTE2(name, __COUNTER__)
   |  |  |  |  ------------------
-  |  |  |  |  |  |    9|      0|#define PASTE2(x, y) PASTE(x, y)
+  |  |  |  |  |  |    9|  18.4E|#define PASTE2(x, y) PASTE(x, y)
   |  |  |  |  |  |  ------------------
-  |  |  |  |  |  |  |  |    8|      0|#define PASTE(x, y) x ## y
+  |  |  |  |  |  |  |  |    8|  18.4E|#define PASTE(x, y) x ## y
   |  |  |  |  |  |  ------------------
   |  |  |  |  ------------------
   |  |  ------------------
   ------------------
-   60|      0|                f();
-   61|      0|            }
-   62|      0|        } catch (...) {
+   60|  18.4E|                f();
+   61|  18.4E|            }
+   62|  18.4E|        } catch (...) {
    63|      0|            --nThreadsServicingQueue;
    64|      0|            throw;
    65|      0|        }
-   66|      1|    }
+   66|      0|    }
    67|      1|    --nThreadsServicingQueue;
    68|      1|    newTaskScheduled.notify_one();
    69|      1|}
@@ -169245,11 +169242,11 @@
   108|       |    int nThreadsServicingQueue GUARDED_BY(newTaskMutex){0};
   109|       |    bool stopRequested GUARDED_BY(newTaskMutex){false};
   110|       |    bool stopWhenEmpty GUARDED_BY(newTaskMutex){false};
-  111|      6|    bool shouldStop() const EXCLUSIVE_LOCKS_REQUIRED(newTaskMutex) { return stopRequested || (stopWhenEmpty && taskQueue.empty()); }
-                                                                                                           ^2^2               ^0
+  111|      4|    bool shouldStop() const EXCLUSIVE_LOCKS_REQUIRED(newTaskMutex) { return stopRequested || (stopWhenEmpty && taskQueue.empty()); }
+                                                                                                           ^0^0               ^0
   ------------------
-  |  Branch (111:77): [True: 4, False: 2]
-  |  Branch (111:95): [True: 0, False: 2]
+  |  Branch (111:77): [True: 4, False: 0]
+  |  Branch (111:95): [True: 0, False: 0]
   |  Branch (111:112): [True: 0, False: 0]
   ------------------
   112|       |};
@@ -227794,7 +227791,7 @@
    75|  1.25k|inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, MutexType* cs, bool fTry = false) {}
   ------------------
   | void EnterCritical<std::mutex>(char const*, char const*, int, std::mutex*, bool):
-  |   75|    943|inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, MutexType* cs, bool fTry = false) {}
+  |   75|    942|inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, MutexType* cs, bool fTry = false) {}
   ------------------
   | void EnterCritical<std::recursive_mutex>(char const*, char const*, int, std::recursive_mutex*, bool):
   |   75|    309|inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, MutexType* cs, bool fTry = false) {}
@@ -228041,7 +228038,7 @@
   197|  1.25k|    {
   198|  1.25k|        if (Base::owns_lock())
   ------------------
-  |  Branch (198:13): [True: 943, False: 1]
+  |  Branch (198:13): [True: 943, False: 0]
   ------------------
   |  Branch (198:13): [True: 309, False: 0]
   ------------------
@@ -228051,13 +228048,13 @@
   200|  1.25k|    }
   ------------------
   | UniqueLock<AnnotatedMixin<std::mutex> >::~UniqueLock():
-  |  197|    944|    {
-  |  198|    944|        if (Base::owns_lock())
+  |  197|    943|    {
+  |  198|    943|        if (Base::owns_lock())
   |  ------------------
-  |  |  Branch (198:13): [True: 943, False: 1]
+  |  |  Branch (198:13): [True: 943, False: 0]
   |  ------------------
   |  199|    943|            LeaveCritical();
-  |  200|    944|    }
+  |  200|    943|    }
   ------------------
   | UniqueLock<AnnotatedMixin<std::recursive_mutex> >::~UniqueLock():
   |  197|    309|    {
@@ -228108,20 +228105,15 @@
   | Unexecuted instantiation: UniqueLock<AnnotatedMixin<std::mutex> >::reverse_lock::reverse_lock(UniqueLock<AnnotatedMixin<std::mutex> >&, char const*, char const*, int)
   ------------------
   223|       |
-  224|      1|        ~reverse_lock() {
-  225|      1|            templock.swap(lock);
-  226|      1|            EnterCritical(lockname.c_str(), file.c_str(), line, lock.mutex());
-  227|      1|            lock.lock();
-  228|      1|        }
+  224|      0|        ~reverse_lock() {
+  225|      0|            templock.swap(lock);
+  226|      0|            EnterCritical(lockname.c_str(), file.c_str(), line, lock.mutex());
+  227|      0|            lock.lock();
+  228|      0|        }
   ------------------
   | Unexecuted instantiation: UniqueLock<AnnotatedMixin<std::recursive_mutex> >::reverse_lock::~reverse_lock()
   ------------------
-  | UniqueLock<AnnotatedMixin<std::mutex> >::reverse_lock::~reverse_lock():
-  |  224|      1|        ~reverse_lock() {
-  |  225|      1|            templock.swap(lock);
-  |  226|      1|            EnterCritical(lockname.c_str(), file.c_str(), line, lock.mutex());
-  |  227|      1|            lock.lock();
-  |  228|      1|        }
+  | Unexecuted instantiation: UniqueLock<AnnotatedMixin<std::mutex> >::reverse_lock::~reverse_lock()
   ------------------
   229|       |
   230|       |     private:
@@ -228137,7 +228129,7 @@
   240|       |     friend class reverse_lock;
   241|       |};
   242|       |
-  243|      0|#define REVERSE_LOCK(g) typename std::decay<decltype(g)>::type::reverse_lock UNIQUE_NAME(revlock)(g, #g, __FILE__, __LINE__)
+  243|  18.4E|#define REVERSE_LOCK(g) typename std::decay<decltype(g)>::type::reverse_lock UNIQUE_NAME(revlock)(g, #g, __FILE__, __LINE__)
   244|       |
   245|       |// When locking a Mutex, require negative capability to ensure the lock
   246|       |// is not already held
⚠️

The coverage was not deterministic between runs.
The fuzz target input was ./p2pheaderpresync_newcorp/p2p_headers_presync/30ad165903c86c0b30bd5940c0031c9f89bd264d.
```

</details>

Instability in `scheduler.cpp`?

janb84 commented:
New corpus all runs end up non-deterministic: 

Used command to test: 
```shell
 cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/build_fuzz/ $PWD/qa-assets/fuzz_corpora/ p2p_headers_presync 9
```

Outcome: 
```shell
The coverage was not deterministic between runs.
```

Let me know if you need additional info or I need to test in a different way. (tested 4 times all runs end up like this) 

<details>
<summary>diff</summary>

```
@@ -144603,7 +144603,7 @@
|  185|  12.5k|    RandomNumberGenerator auto& Impl() noexcept { return static_cast<T&>(*this); }
   ------------------
   | _ZN11RandomMixinI21InsecureRandomContextE4ImplEv:
-  |  185|    668|    RandomNumberGenerator auto& Impl() noexcept { return static_cast<T&>(*this); }
+  |  185|    660|    RandomNumberGenerator auto& Impl() noexcept { return static_cast<T&>(*this); }
   ------------------
   186|       |
   187|       |protected:
@@ -145092,12 +145092,12 @@
   ------------------
   | Unexecuted instantiation: _ZN11RandomMixinI21InsecureRandomContextE3maxEv
   ------------------
-  367|    668|    inline uint64_t operator()() noexcept { return Impl().rand64(); }
+  367|    660|    inline uint64_t operator()() noexcept { return Impl().rand64(); }
   ------------------
   | Unexecuted instantiation: _ZN11RandomMixinI17FastRandomContextEclEv
   ------------------
   | _ZN11RandomMixinI21InsecureRandomContextEclEv:
-  |  367|    668|    inline uint64_t operator()() noexcept { return Impl().rand64(); }
+  |  367|    660|    inline uint64_t operator()() noexcept { return Impl().rand64(); }
   ------------------
   368|       |};
   369|       |
@@ -145175,14 +145175,14 @@
   437|      0|    }
   438|       |
   439|       |    constexpr uint64_t rand64() noexcept
-  440|    668|    {
-  441|    668|        uint64_t s0 = m_s0, s1 = m_s1;
-  442|    668|        const uint64_t result = std::rotl(s0 + s1, 17) + s0;
-  443|    668|        s1 ^= s0;
-  444|    668|        m_s0 = std::rotl(s0, 49) ^ s1 ^ (s1 << 21);
-  445|    668|        m_s1 = std::rotl(s1, 28);
-  446|    668|        return result;
-  447|    668|    }
+  440|    660|    {
+  441|    660|        uint64_t s0 = m_s0, s1 = m_s1;
+  442|    660|        const uint64_t result = std::rotl(s0 + s1, 17) + s0;
+  443|    660|        s1 ^= s0;
+  444|    660|        m_s0 = std::rotl(s0, 49) ^ s1 ^ (s1 << 21);
+  445|    660|        m_s1 = std::rotl(s1, 28);
+  446|    660|        return result;
+  447|    660|    }
   448|       |};
   449|       |
   450|       |
⚠️

The coverage was not deterministic between runs.
```

</details>

marcofleon commented:
@janb84 I could be wrong but I think that might be coming from faaf3819f2c957a4c9688386b8301a1ef70ace07. If so, then https://github.com/bitcoin/bitcoin/pull/32198#issuecomment-2783187937 mentioned it as a source of non-determinism to ignore for now.

janb84 reviewed (commented):
tACK [fa2d344](https://github.com/bitcoin/bitcoin/commit/fa2d3441c6557adba3d393abfc1af22fb788b5c2)

> @janb84 I could be wrong but I think that might be coming from [faaf381](https://github.com/bitcoin/bitcoin/commit/faaf3819f2c957a4c9688386b8301a1ef70ace07). If so, then [#32198 (comment)](https://github.com/bitcoin/bitcoin/pull/32198#issuecomment-2783187937) mentioned it as a source of non-determinism to ignore for now.

Ah my bad, seems logical. Thanks @marcofleon for pointing this out to me, somehow I didn't made the connection between the two.

Commit:
fuzz: Disable unused validation interface and scheduler in p2p_headers_presync

This may also avoid non-determinism in the scheduler thread.

janb84 reviewed (commented):
Re-ACK [faee556](https://github.com/bitcoin/bitcoin/pull/32198/commits/faee5569c0797084ac74e0173c1969d6108b180c)

Change since last ACK:
- removed he scheduler functionality 

(I'm too junior PR reviewer to judge if the randomness issue has to be solved in this PR)

Commit:
fuzz: Avoid influence on the global RNG from peerman m_rng

This should avoid the remaining non-determistic code coverage paths.

Without this patch, the tool would report a diff (only when running
without libFuzzer):

cargo run --manifest-path ./contrib/devtools/deterministic-fuzz-coverage/Cargo.toml -- $PWD/bld-cmake/ $PWD/../qa-assets/fuzz_corpora/ p2p_headers_presync 32

janb84 reviewed (commented):
Re-ACK [fa3cd7a](https://github.com/bitcoin/bitcoin/pull/32198/commits/fa3cd7aac74acf576a85299c024249c81384b921)

When replacing `std::ranges::shuffle(files, InsecureRandomContext{uint64_t(SystemClock::now().time_since_epoch().count())});` with `std::ranges::shuffle(files, std::mt19937{std::random_device{}()});` in this commit as suggested,  the test is deterministic ! 

 Have run the fuzz test 10x ✅

janb84 reviewed (commented):
Re-ACK [faa3ce3](https://github.com/bitcoin/bitcoin/commit/faa3ce31998cdbdcb888132b0d38ee7a1e743682)

Test is deterministic without changes ✅ (tested 10x) 

> > When replacing `std::ranges::shuffle(files, InsecureRandomContext{uint64_t(SystemClock::now().time_since_epoch().count())});` with `std::ranges::shuffle(files, std::mt19937{std::random_device{}()});` in this commit as suggested, the test is deterministic !
> 
> Ok, now I see that this is required for libc++, which I guess you are using? (I was using the GCC std lib)

Yes the macos version of it & Clang 19

> 
> Pushed that diff as well.
> 
> Hopefully this is the last one 😅

I'm sorry, I was under the impression you wanted to do that change in a later PR.

marcofleon reviewed (commented):
ACK faa3ce31998cdbdcb888132b0d38ee7a1e743682

[coverage on master](https://marcofleon.github.io/coverage/p2pheaderspresync_master/)
[coverage on this PR](https://marcofleon.github.io/coverage/p2pheaderpresync_32198/)

Coverage on this PR is just from a run overnight but seems to be about the same where it counts.

I've run the script 10 times or so and the determinism is looking good!

Crypt-iQ commented:
tACK faf4c1b6fc330885ddf84b643838d1e301aaeab2

Have verified that the shuffle change in the first commit gets rid of the non-determinism.  Just a question, does disabling the scheduler remove some important coverage that otherwise wouldn't be hit?

Crypt-iQ commented:
tACK faa3ce31998cdbdcb888132b0d38ee7a1e743682

> fyi, looks like your review is stale (not the top commit hash)

(Sorry, copied the wrong commit hash when making the comment)
